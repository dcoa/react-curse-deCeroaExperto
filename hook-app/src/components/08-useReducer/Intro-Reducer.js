const initialState =[{
    id: 1,
    todo: 'Hacer ejercicio',
    done: false
}]

const todoReducer = ( state = initialState, action) => {
//si accion existe busque la kye type
    if(action?.type === 'add') {
        return [...state, action.payload]
    }
    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Concentir Luna',
    done: false
}

//Type es la accion a dedarrollar en el reducer
//Payload va a hacer aquello por lo cual se va a ejecuar la accion, la data
const actionAddNewToDo = {
    type: 'add',
    payload: newTodo
}
todos = todoReducer(todos, actionAddNewToDo);
console.log(todos)