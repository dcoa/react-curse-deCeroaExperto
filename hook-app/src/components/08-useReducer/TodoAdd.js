import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAdd }) => {
    const { formValues, handleInputChange, reset } = useForm({
        description: ''
    })

    const handleSubmit = (e) => {
        e.preventDefault();
        if (formValues.description.trim() === '') return;
        const newTodo = {
            id: new Date().getTime(),
            desc: formValues.description,
            done: false
        }
        handleAdd(newTodo)
        reset()
    }
    return (
        <>
            <h4>Agregar nueva tarea</h4>
            <form className='d-grid gap-2 d-md-block'
                onSubmit={handleSubmit}
            >
                <input
                    type="text"
                    name="description"
                    placeholder="Nueva tarea"
                    autoComplete="off"
                    className='form-control'
                    onChange={handleInputChange}
                    value={formValues.description}
                />
                <button
                    type='submit'
                    className='btn btn-outline-success mt-2'                        >
                    Agregar
                </button>

            </form>
        </>
    )
}
