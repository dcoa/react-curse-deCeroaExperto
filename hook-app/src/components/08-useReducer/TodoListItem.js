import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = ({ todo, index, handleToggle, handleDelete }) => {
    return (
        <li className='list-group-item'>
            <p className={`${todo.done && 'doneTask'}`}
                onClick={() => handleToggle(todo.id)}
            >
                {index + 1}. {todo.desc}
            </p>
            <button
                className='btn btn-sm btn-danger'
                onClick={() => handleDelete(todo.id)}
            >Borrar</button>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}
