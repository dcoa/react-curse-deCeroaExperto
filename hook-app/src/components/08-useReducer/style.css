body {
    padding: 50px;
}

.list-group-item {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.list-group-item p {
    margin: 0;
}

.doneTask {
    text-decoration: line-through;
}