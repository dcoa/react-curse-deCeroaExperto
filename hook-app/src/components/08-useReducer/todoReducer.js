export const todoReducer = ( state = [], action) => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload]   
            case 'delete':
                return state.filter(ob => ob.id !== action.payload) //payload es el id 
                case 'toggle':
                    return state.map(ob => ob.id === action.payload ? {...ob, done: !ob.done} : ob) //payload es el id
        default:
            return state;
    }
}