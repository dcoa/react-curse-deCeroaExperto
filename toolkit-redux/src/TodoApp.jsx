import { useState } from "react"
import { useGetTodoByIdQuery, useGetTodosQuery } from "./store/api/todosApi"


export const TodoApp = () => {
  // const {data: todos = [], isLoading } = useGetTodosQuery()
  const [todoId, setTodoId] = useState(1)
  const { data: todo, isLoading } = useGetTodoByIdQuery(todoId)

  const nextTodo = () => {
    setTodoId( todoId + 1 )
  }
  const prevTodo = () => {
    if (todoId === 1) return
    setTodoId( todoId - 1 )
  }
  return (
    <>
      <h1>Todos - RTK query</h1>
      <pre>...</pre>
      <h4>isLoading: {isLoading ? 'true' : 'false'}</h4>
      <pre>...</pre>

      {/* <ul>
      {
        todos.map(todo => (<li key={todo.id}>{todo.title} <strong>{todo.completed ? 'completed' : 'no yet'}</strong></li>))
      }
    </ul> */}
      <button onClick={prevTodo}>Prev ToDo</button>

      <button onClick={nextTodo}>Next ToDo</button>
      <br />
      <code>{JSON.stringify(todo)}</code>
    </>
  )
}
