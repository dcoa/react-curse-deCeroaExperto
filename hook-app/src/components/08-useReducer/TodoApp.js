import React, { useEffect, useReducer } from 'react'
import './style.css'
import { TodoAdd } from './TodoAdd'
import { TodoList } from './TodoList'
import { todoReducer } from './todoReducer'

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'Aprender react',
//     done: false
// }]

const init = () => {
    // utilizo or para el retorno evitar el null
    return   JSON.parse(localStorage.getItem('todos')) || []
}
export const TodoApp = () => {
    // init es una funcion que permite declarar el estado inicial de acuerdo con lo que haga
    const [todos, dispatch] = useReducer(todoReducer, [], init);
    //useReducer sin init
    // const [todos, dispatch] = useReducer(todoReducer, initialState,);


    useEffect(()=>{
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);

    const handleAdd = (newTodo) => {
        // es al que se le envia la accion
        dispatch({
            type: 'add',
            payload: newTodo
        })

    }
    const handleDelete = (todoid) => {
        const acctionTodoDelete = {
            type: 'delete',
            payload: todoid
        }
        dispatch(acctionTodoDelete)

    }

    const handleToggle = (todoid)=> {
        dispatch({
            type: 'toggle',
            payload: todoid
        })
    }
    return (
        <div >
            <h1>To Do App: {todos.length}</h1>

            <div className='row'>
                <div className='col-5'>
                    <TodoList 
                    todos={todos} 
                    handleDelete={handleDelete} 
                    handleToggle={handleToggle}/>
                </div>
                <div className='col-5' >
                    <TodoAdd handleAdd={handleAdd}/>
                </div>
            </div>
        </div>
    )
}
