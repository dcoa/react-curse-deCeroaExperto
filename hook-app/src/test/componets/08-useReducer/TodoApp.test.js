import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";

describe('Test on TodoApp component', () => {

    const wrapper = shallow(<TodoApp/>)
    
    Storage.prototype.setItem = jest.fn(() => {});
    test('should show the componet correctly', () => {

        expect(wrapper).toMatchSnapshot();
        
    });

    test('should add a new task', () => {
        // probar el componete app en general
        const wrapper = mount(<TodoApp/>);
        act(()=>{
            wrapper.find('TodoAdd').prop('handleAdd')( demoTodos[0] );
            wrapper.find('TodoAdd').prop('handleAdd')( demoTodos[1] );
        });


        expect(wrapper.find('h1').text().trim()).toBe('To Do App: 2');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2)

    });


    test('should delete a todo', () => {
        wrapper.find('TodoAdd').prop('handleAdd')( demoTodos[0] );

        expect(wrapper.find('h1').text().trim()).toBe('To Do App: 1');
        wrapper.find('TodoList').prop('handleDelete')( demoTodos[0].id );
        expect(wrapper.find('h1').text().trim()).toBe('To Do App: 0');


    })
    
    
    

});
