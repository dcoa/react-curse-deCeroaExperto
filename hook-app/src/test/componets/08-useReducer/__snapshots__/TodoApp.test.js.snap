// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on TodoApp component should show the componet correctly 1`] = `
<div>
  <h1>
    To Do App: 
    0
  </h1>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
